openhcl/virt_mshv_vtl/src/processor/tdx/mod.rs

```diff
@@ -2084,78 +2102,97 @@ impl UhProcessor<'_, TdxBacked> {
     }
 }
 
-impl<T: CpuIo> EmulatorSupport for UhEmulationState<'_, '_, T, TdxBacked> {
+impl<T: CpuIo> X86EmulatorSupport for UhEmulationState<'_, '_, T, TdxBacked> {
     type Error = UhRunVpError;
 
     fn vp_index(&self) -> VpIndex {
         self.vp.vp_index()
     }
 
-    fn vendor(&self) -> x86defs::cpuid::Vendor {
-        self.vp.partition.caps.vendor
-    }
-
-    fn state(&mut self) -> Result<x86emu::CpuState, Self::Error> {
-        let cr0 = self.vp.backing.cr0.read(&self.vp.runner);
-        let efer = self.vp.backing.efer;
+    fn load_registers(&mut self) {
```

Comment:
I think I'd like to take this in a slightly different direction, namely getting rid of load_registers on this trait entirely. If something is cheap to access then we can just access it directly in the getter and setter methods. If something is expensive to access then that's where the cache comes in. That way we can ensure we're performing any expensive operations at most once throughout an entire instruction, but also allows us to skip performing it at all if it's not necessary. Right now load_registers is doing a lot of work up front that we could be doing lazilly instead, and I think moving to that more lazy model is the ultimate goal here.

Reply:
This makes a lot of sense to me, I have implemented for TDX/SNP. I will likely kill the flush as well, as none of the expensive registers are mutable (or at least, there is not yet a case where the emulator modifies them)

However, for MSHV, it looks to me like preloading registers could actually be useful. Currently, MSHV queries a bunch of registers in batch via ioctl to fill the CpuState struct. If we switch to individually retrieving those registers lazily, would this not result in a bunch of extra ioctl overhead?

Reply:
That's a good point. Normally I'd say we should do that lazily, but we're always going to be accessing EFER, CR0 and CS to do the bitness check. So I guess being eager for MSHV makes more sense. I'm really not a fan of having load_registers on the trait though, as that leaves the possibility open of forgetting to call it. Instead I'm thinking we should remove the Default bound on the EmulationCache associated type and have each backing pass in a cache to the emulate call. SNP and TDX (and MSHV ARM) can just pass in Default::default(), but MSHV x64 can construct a type with all these registers pre-filled and just make the single ioctl to construct it.

Reply:
That sounds good to me, implemented

FWIW I have a similar concern about forgetting to flush the cache, but that is at least not an issue introduced by this refactor, flushing the state was a necessity before the cache. I moved the cache flushes back to the same locations that we were flushing `state` previously.